### ISSUE TITLE
Shouldn't close on the Sink return Async::NotReady until the remote has responded with a close frame?

### ISSUE THREAD

**Issue by najamelan** [snapview/tokio-tungstenite#69]
Seems to have been somewhat discussed in #4. 

`close` being async, it seems to me that it should only resolve once the remote has confirmed the close. This allows client code to await the close before dropping underlying network connection in order to close down properly.

In an async-await version this would imply storing the waker and waking up a task waiting for the close to complete.

**Comment by najamelan** [comment]
From trying this out in the library I use tokio-tungstenite from, I notice the following. If the remote end only tries to send, and doesn't do a read, it will never see the close frame, and thus never answer it. 

The code that triggers this is test code. In real life code there is almost always a reader task which will process incoming messages in a loop, triggering a read wakeup when a new message arrives, but it does mean that with such an implementation of close, it should always be combined with a timeout, just in case, to avoid the program hanging.

**Comment by najamelan** [comment]
Playing around with this a bit more, I notice that it's not possible to receive incoming messages anymore after sending a close frame. 

Since the remote might be sending out a data message just when you send a close, it seems like the receiving end should remain functional until we receive an acknowledgement. That can avoid data loss. Since the sending on the remote did not return an error, and it happened before they acknowledged a close frame, it seems to me that tungstenite should not silently force data loss on the user.

I can't really see a good reason to swallow incoming data whilst waiting for the close to be acknowledged. Client code can choose to ignore those messages if they want, but tungstenite shouldn't make that decision.

**Comment by najamelan** [comment]
@application-developer-DA Ok, so I dug some more through tungstenite source code and so I now see that it handles keeping the underlying connection alive long enough by waiting to send you `Error::ConnectionClosed` until it is safe to drop the underlying connection. That implementation has some issues of it's own, as described in snapview/tungstenite-rs#77, but all in all it does seem like pending the call to close until then is more hassle than it's worth.

I did a writeup of what I found out so others can benefit from documentation rather than having to dig the source code: snapview/tungstenite-rs#79.
